crates/ol/state-support-types/src/batch_diff_layer.rs

```diff
@@ -0,0 +1,544 @@
+//! Read-only state accessor that overlays WriteBatch diffs on a base state.
+//!
+//! This provides an `IStateAccessor` implementation that checks a stack of
+//! `WriteBatch` references before falling back to a base state. All write
+//! operations are unsupported since this is read-only.
+
+use strata_acct_types::{AccountId, AccountSerial, AcctError, AcctResult, BitcoinAmount};
+use strata_asm_manifest_types::AsmManifest;
+use strata_identifiers::{Buf32, EpochCommitment, L1BlockId, L1Height};
+use strata_ledger_types::{IStateAccessor, NewAccountData};
+
+use crate::write_batch::WriteBatch;
+
+/// A read-only state accessor that overlays a stack of WriteBatch diffs.
+///
+/// Reads check each batch in reverse order (last = most recent), then fall back
+/// to the base state. All write operations return `AcctError::Unsupported` or
+/// silently no-op (for setters that return `()`).
+///
+/// The batch slice can be empty, making this a read-only wrapper for the base.
+/// This is useful for scenarios where you want to view state with pending
+/// changes applied without modifying anything.
+pub struct BatchDiffState<'batches, 'base, S: IStateAccessor> {
+    base: &'base S,
+    batches: &'batches [WriteBatch<S::AccountState>],
```

Comment:
Seems like it is possible to have a merged batch here. We would merge by consuming an earlier batch to avoid clones.
This way we don't have to linearly scan through batches to find accounts. Especially when the id look up is also linear.